### PR TITLE
docs: move docs from juju.is

### PR DESCRIPTION
**Pull request by tmihoc**
As part of a broader move where juju.is docs go federated and RTD, this PR adds all the python-libjuju material we previously had under juju.is/docs/juju (in how-to guide tabs) to the python-libjuju RTD (under a new section called "How-to guides").

This RTD project could use a lot of further improvements (e.g., for all of the other projects -- juju, terraform-provider-juju, charmcraft, ops -- I made sure to update the homepage according to the standard template and to have the full Diataxis tree and integrate the new content nicely with the old, removing duplication, etc.). However, given that we are slowly phasing python-libjuju out, I believe leaving things as is is probably fine.

There is one fix that it is worth making, namely, the links to Juju docs. I followed the instructions in https://canonical-documentation-with-sphinx-and-readthedocscom.readthedocs-hosted.com/style-guide/#internal-references for "targets in other doc sets", but the links don't seem to be working. It could be because the Juju RTD project needs some fixing (we're working on it). I should know soon enough what needs to be done and then I'll come back and fix all the links in a follow-up PR.

#### QA Steps

From the root of the repo:

```
# Build the docs:
tox -e docs 

#Preview the docs:
open docs/_build/index.html
```

### CODE REVIEW

File: docs/howto/#manage-machines.rst# [hunk 1]
```diff
@@ -0,0 +1,161 @@
+.. _manage-machines:
```
Comment (benhoyt) [inline docs/howto/#manage-machines.rst#]: Is this `#manage-machines.rst#` file an editor backup? Presumably it shouldn't be committed.

**Review by benhoyt — CHANGES_REQUESTED** [review]
Thanks for this, @tmihoc! However, it looks like you've accidentally commits all the Python files under `build/*`. Where are these files coming from? Should the `/build` directory be added to `.gitignore`?

**Comment by james-garner-canonical** [comment]
I've taken care of rebasing this branch, and added a commit that removes `build/` and `#manage-machines.rst#`, updates `.gitignore` to ignore `build/`, and adds the (mostly whitespace) changes mandated by `pre-commit`.

**Review by tonyandrewmeyer — APPROVED** [review]
This makes sense to me. The how-to guides are integrated, which is the main point, and while the docs site is still not great, I agree that it doesn't make sense to fix that given where things are headed.

For the broken Juju docs, the plan to fix them in a future PR seems fine - I don't think it needs to block merging, even though we did that that approach with ops. For what it's worth, I don't see the `juju` cross-Sphinx link defined, so that may be the problem?

File: docs/howto/manage-applications.rst [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+.. code:: python
+
+   await my_model.deploy(..., constraints={, 'arch': 'amd64', 'mem': 256}, ...)
+
+
+* To set constraints post deployment, you may use the `Application.set_contraints()` method, similar to passing constraints in the deploy call above:
+
+.. code:: python
+
+   await my_app.set_constraints(constraints={, 'arch': 'amd64', 'mem': 256})
```
Comment (dimaqq) [inline docs/howto/manage-applications.rst]:
```suggestion
   await my_app.set_constraints(constraints={'arch': 'amd64', 'mem': 256})
```

File: docs/howto/manage-applications.rst [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+> See also: :ref:`juju:constraint`
+
+**Set values.** To set constraints for application in python-libjuju:
+
+* To set at deployment, simply provide a `constraints` map during the `Model.deploy()` call:
+
+.. code:: python
+
+   await my_model.deploy(..., constraints={, 'arch': 'amd64', 'mem': 256}, ...)
```
Comment (dimaqq) [inline docs/howto/manage-applications.rst]:
```suggestion
   await my_model.deploy(..., constraints={'arch': 'amd64', 'mem': 256}, ...)
```

File: docs/howto/manage-storage-pools.rst [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+Update a storage pool
+---------------------
+
+To update an existing storage pool attributes, on a connected Model object, use the `update_storage_pool()` method, passing the name of the storage and the attribute values to update. For example:
+
+.. code:: python
+
+   await my_model.update_storage_pool(
+       "operator-storage",
+       attributes={"volume-type":"provisioned-iops", "iops"="40"})
```
Comment (dimaqq) [inline docs/howto/manage-storage-pools.rst]:
```suggestion
       attributes={"volume-type": "provisioned-iops", "iops": "40"})
```

File: docs/howto/manage-credentials.rst [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+-------------------
+
+To update a credential, on a connected `Controller` object, use the `Controller.add_credential()` method. `add_credential` is an upsert method (where it inserts if the given credential is new, and updates if the given credential name already exists).
+
+.. code:: python
+
+   from juju.client import client as jujuclient
+
+   my_controller.add_credential("my-credential",
+       jujuclient.CloudCredential(auth_type="jsonfile", attrs={'file':'path_to_cred_file'})
```
Comment (dimaqq) [inline docs/howto/manage-credentials.rst]:
```suggestion
   my_controller.add_credential("my-credential",
       jujuclient.CloudCredential(auth_type="jsonfile", attrs={'file':'path_to_cred_file'}))
```

File: docs/howto/manage-ssh-keys.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+To add a public `ssh` key to a model, on a connected Model object, use the `add_ssh_key()` method, passing a name for the key and the actual key payload. For example:
+
+.. code:: python
+
+   SSH_KEY = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC1I8QDP79MaHEIAlfh933zqcE8LyUt9doytF3YySBUDWippk8MAaKAJJtNb+Qsi+Kx/RsSY02VxMy9xRTp9d/Vr+U5BctKqhqf3ZkJdTIcy+z4hYpFS8A4bECJFHOnKIekIHD9glHkqzS5Vm6E4g/KMNkKylHKlDXOafhNZAiJ1ynxaZIuedrceFJNC47HnocQEtusPKpR09HGXXYhKMEubgF5tsTO4ks6pplMPvbdjxYcVOg4Wv0N/LJ4ffAucG9edMcKOTnKqZycqqZPE6KsTpSZMJi2Kl3mBrJE7JbR1YMlNwG6NlUIdIqVoTLZgLsTEkHqWi6OExykbVTqFuoWJJY3BmRAcP9H3FdLYbqcajfWshwvPM2AmYb8V3zBvzEKL1rpvG26fd3kGhk3Vu07qAUhHLMi3P0McEky4cLiEWgI7UyHFLI2yMRZgz23UUtxhRSkvCJagRlVG/s4yoylzBQJir8G3qmb36WjBXxpqAGHfLxw05EQI1JGV3ReYOs= user@somewhere"
+   await my_model.add_ssh_key('admin', SSH_KEY)
+
+
+> See more: `add_ssh_key() `https://pythonlibjuju.readthedocs.io/en/latest/api/juju.model.html#juju.model.Model.add_ssh_key>`_, `Model (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/model.html>`_
```
Comment (dimaqq) [inline docs/howto/manage-ssh-keys.rst]:
```suggestion
> See more: `add_ssh_key() https://pythonlibjuju.readthedocs.io/en/latest/api/juju.model.html#juju.model.Model.add_ssh_key>`_, `Model (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/model.html>`_
```

File: docs/howto/manage-storage.rst [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+--------------
+
+To detach a storage instance from a unit, on a Unit object, use the `detach_storage()` method, passing the storage id as an argument. For example:
+
+.. code:: python
+
+   await my_unit.detach_storage("osd-devices/2")
+
+
+> See more: `detach_storage() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.unit.html#juju.unit.Unit.detach_storage>`_, `Unit (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/unit.html)>`_
```
Comment (dimaqq) [inline docs/howto/manage-storage.rst]:
```suggestion
> See more: `detach_storage() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.unit.html#juju.unit.Unit.detach_storage>`_, `Unit (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/unit.html>`_
```

File: docs/howto/manage-secret-backends.rst [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+Add a secret backend to a model
+-------------------------------
+
+To add a secret backend to a controller, on a connected Controller, use the `add_secret_backends()` method, passing the `id`, `name`, `backend_type`, and `config` as arguments. For example:
+
+.. code:: python
+
+   await my_controller.add_secret_backends("1001", "myvault", "vault", {"endpoint": vault_url, "token": keys["root_token"]})
+
+> See more: `add_secret_backend() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.controller.html#juju.controller.Controller.add_secret_backends>`_, `Controller (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/controller.html>`_
```
Comment (dimaqq) [inline docs/howto/manage-secret-backends.rst]:
```suggestion
> See more: `add_secret_backends() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.controller.html#juju.controller.Controller.add_secret_backends>`_, `Controller (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/controller.html>`_
```

File: docs/howto/manage-secret-backends.rst [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+.. code:: python
+
+   await my_controller.update_secret_backends(
+               "myvault",
+               name_change="changed_name")
+
+Check out the documentation for the full list of arguments.
+
+> See more: `update_secret_backend() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.controller.html#juju.controller.Controller.update_secret_backends>`_, `Controller (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/controller.html>`_
```
Comment (dimaqq) [inline docs/howto/manage-secret-backends.rst]:
```suggestion
> See more: `update_secret_backends() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.controller.html#juju.controller.Controller.update_secret_backends>`_, `Controller (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/controller.html>`_
```

File: docs/howto/manage-units.rst [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+   my_app.scale(4)
+
+
+> See more: `Application (module) <https://pythonlibjuju.readthedocs.io/en/latest/narrative/application.html>`_, `Application.add_unit() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.application.html#juju.application.Application.add_unit>`_, `Application.scale() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.application.html#juju.application.Application.scale>`_, `Application.destroy_units() <https://pythonlibjuju.readthedocs.io/en/latest/api/juju.application.html#juju.application.Application.destroy_units>`_
+
+
+Show details about a unit
+-------------------------
+
+Too see details about a unit in `python-libjuju` client, you can use various fields and methods of a `Unit` object. For example, to get the `public_address` of a unit:
```
Comment (dimaqq) [inline docs/howto/manage-units.rst]:
```suggestion
To see details about a unit in `python-libjuju` client, you can use various fields and methods of a `Unit` object. For example, to get the `public_address` of a unit:
```

File: docs/howto/manage-storage.rst [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+Detach storage
+--------------
+
+To detach a storage instance from a unit, on a Unit object, use the `detach_storage()` method, passing the storage id as an argument. For example:
+
+.. code:: python
+
+   await my_unit.detach_storage("osd-devices/2")
+
+
```
Comment (dimaqq) [inline docs/howto/manage-storage.rst]:
```suggestion
   await my_unit.detach_storage("osd-devices/2")

```

File: docs/howto/manage-units.rst [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+Mark unit errors as resolved
+----------------------------
+
+To mark unit errors as resolved in the `python-libjuju` client, you can call the `resolved()` method on a `Unit` object:
+
+.. code:: python
+
+   my_unit.resolved()
+
+> See more: `Unit.resolved()`
```
Comment (dimaqq) [inline docs/howto/manage-units.rst]:
```suggestion
   await my_unit.resolved()

> See more: `Unit.resolved() https://pythonlibjuju.readthedocs.io/en/latest/api/juju.unit.html#juju.unit.Unit.resolved`_
```

File: docs/howto/manage-users.rst [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+To enable or disable a user, on a User object, use the `enable()` and `disable()` methods.
+
+.. code:: python
+
+   await user_object.enable()
+
+   await user_object.disable()
+
+You can also check if a user is enabled or disabled using the `enabled` and `disabled` properties on the Unit object.
```
Comment (dimaqq) [inline docs/howto/manage-users.rst]:
```suggestion
You can also check if a user is enabled or disabled using the `enabled` and `disabled` properties on the User object.
```

**Review by dimaqq — COMMENTED** [review]
first read

**Comment by tmihoc** [comment]
@james-garner-canonical I see you've already committed some improvements. Do you think I could leave this PR in your hands? Asking because I'm juggling changes in a few other repos at the moment and I could use all the help I can get.

**Comment by james-garner-canonical** [comment]
I've merged @dimaqq's suggestions and signed the commit.

Re: linking to Juju docs, @tonyandrewmeyer's comment pointed me to us needing to define the `juju:` reference for sphinx. [intersphinx](https://docs.readthedocs.io/en/stable/guides/intersphinx.html) is already in `docs/conf.py`'s `extensions`, so I guess we need something like this -- but what url do we use here? http://canonical-juju.readthedocs-hosted.com/

```py
intersphinx_mapping = {
    'python': ('https://docs.python.org/3', None),
    "juju": (JUJU_DOCS_URL, None),
}
```

**Comment by james-garner-canonical** [comment]
I guess if we go with @tmihoc's original suggestion to fix the Juju links in a subsequent PR, then this might be ready to merge. Any objections, @benhoyt and @dimaqq ?

**Review by benhoyt — APPROVED** [review]
Yep, I'm good with it.

The only minor thing I noticed when viewing locally is the breadcrumbs at the top of the page have a "<no-title>" in them:
Screenshot from 2024-12-20 11-21-20

**Pull request merged**